File: ocaml-src/link.ml
open Printf

type link = 
  { tag : string
  ; len : int
  ; uri_len : int
  ; cipher_name_len : int
  ; decryption_key_len : int
  ; uri : string 
  ; cipher : string
  ; decryption_key : string }

let unpack_link bits link =
  bitmatch bits with
  | { tag : 54 ;                            (* Link Tag *)
      len : 8 ;                             (* Link Length *)
      uri_len : 8 ;                         (* URI Length *)
      cipher_name_len : 8 ;                 (* Cipher Name Length *)
      decryption_key_len : 16 ;             (* Decryption Key Length *)
      uri : uri_len ;                       (* URI Value *)
      cipher : cipher_name_len ;            (* Cipher Name *)
      decryption_key : decryption_key_len } (* Decryption key*)
  -> link
  | { _ } as unmatched -> 
      eprintf "Link was unmatched:\n";
      Bitstring.hexdump_bitstring stderr unmatched;
      exit 1

File: data/link.go
package data

/*
	- Encoding?
*/

type Link struct {
	/*
	   magic - 54
	   length - 16
	   uri_length - 8
	   cipher_name_len - 8
	   decryption_key_len - 16
	   uri - uri_len
	   cipher - cipher_name_len
	   decryption_key - decryption_key_len
	*/
	magic              [6]byte
	length             int16
	uri_length         int8
	cipher_name_len    int8
	decryption_key_len int16
	uri                []byte
	cipher             []byte
	decryption_key     []byte
}

func (lnk *Link) create(name []byte, key []byte, cipher_name string) {

	lnk.magic = [6]byte{'L', 'I', 'N', 'K', '\r', '\n'}

	lnk.cipher = []byte(cipher_name)
	lnk.cipher_name_len = int8(len(lnk.cipher))

	lnk.uri = name
	lnk.uri_length = int8(len(lnk.uri))

	lnk.decryption_key = key
	lnk.decryption_key_len = int16(len(lnk.decryption_key))

	lnk.length = 0
}
